Redis-py 8.0.0 - Cannot enable maintenance notifications for connection object that doesn't have a host attribute.

Greetings all, 

In hopes this isn't a redundant issue where I've failed to do a proper amount of research first, I've installed `8.0.0` with the use of `unix_socket_path`. This threw `ValueError: Cannot enable maintenance notifications for connection object that doesn't have a host attribute.`, thus I added `maint_notifications_config=MaintNotificationsConfig(enabled=False)` to the Redis connection object's initialization. It seems like this may be ignored downstream, as it continues to throw the ValueError. 

It seems like the ValueError's raised by PUBSUB, stemming from the `subscribe()` function in `redis/client.py`, which calls `connect_check_health()` downstream, which calls `activate_maint_notifications_handling_if_enabled(check_health=check_health)`, leading to `self._enable_maintenance_notifications()` throwing the ValueError. 

I'm running with `decode_responses` and `retry_on_timeout` enabled, and `legacy_responses` disabled. Have I missed something here, or is this a bug? In my mind, I've missed something, but figured I'd ask here as I prepare to dig further. 

I've also reviewed the 8.0.0 changelog, though didn't see anything that specifically stood out related to this. It _is_ possible I've overlooked something. 

Thank you.

Some of our tests also started to fail with the same error after upgrading `redis-py` to `8.0.0`, see [tests output logs](https://jenkins.softwareheritage.org/job/DENV/job/tests/3310/artifact/swh-indexer.log).

I could bisect the commit that introduced that regression to 1bc0c208e8f0eba0e05b2ea41e5d16d1049144e9 so indeed this is due to changing default RESP protocol version from 2 to 3.

Using [pytest-redis](https://github.com/dbfixtures/pytest-redis), the issue can be reproduced with that simple test file:
```python
def test_redis_flush(redisdb):
    redisdb.flushdb()
    assert True
```

Same error is triggered by some of the `pytest-redis` tests in the dependabot pull request to upgrade redis from 7.4.0 to 8.0.0: https://github.com/dbfixtures/pytest-redis/pull/995.

Digging a bit in the `redis-py` code, that bug is triggered when using unix socket connections instead of TCP ones. Those are used when providing the `unix_socket_path` parameter to the `Redis` client.

`pytest-redis` uses unix socket connections to setup temporary redis cluster as detailed in that failing test report:
```
____________________ ERROR at teardown of test_second_redis ____________________

request = <SubRequest 'redis_otherdb' for <Function test_second_redis>>

    @pytest.fixture
    def redisdb_factory(request: FixtureRequest) -> Generator[redis.Redis, None, None]:
        """Create connection for pytest-redis.
    
        #. Load required process fixture.
        #. Get redis module and config.
        #. Connect to redis.
        #. Flush database after tests.
    
        :param FixtureRequest request: fixture request object
        :rtype: redis.client.Redis
        :returns: Redis client
        """
        proc_fixture: NoopRedis | RedisExecutor = request.getfixturevalue(process_fixture_name)
        config = get_config(request)
    
        redis_host = proc_fixture.host
        redis_port = proc_fixture.port
        redis_username = proc_fixture.username
        redis_password = proc_fixture.password
        redis_db = dbnum
        decode_responses: Literal[True] | Literal[False] = (
            decode if decode is not None else config.decode
        )
    
        redis_client = redis.Redis(
            redis_host,
            redis_port,
            redis_db,
            username=redis_username,
            ***
            unix_socket_path=proc_fixture.unixsocket,
            decode_responses=decode_responses,
        )
    
        yield redis_client
>       redis_client.flushall()

config     = RedisConfig(host='127.0.0.1', port=None, port_search_count=5, username=None, *** exec='/home/runner/redis/r...ce', db_count=8, save=None, compression=False, rdbchecksum=False, syslog=False, decode=False, datadir=None, modules=[])
dbnum      = 0
decode     = None
decode_responses = False
proc_fixture = <pytest_redis.executor.process.RedisExecutor: "/home/runn..." 0x7f3b41da5040>
process_fixture_name = 'redis_other_proc'
redis_client = <redis.client.Redis(<redis.connection.ConnectionPool(<redis.connection.UnixDomainSocketConnection(db=0,username=<REDAC...elaxed_timeout=10, endpoint_type=None),orig_host_address=None,orig_socket_timeout=5,orig_socket_connect_timeout=5)>)>)>
redis_db   = 0
redis_host = '127.0.0.1'
redis_password = None
redis_port = 29995
redis_username = None
request    = <SubRequest 'redis_otherdb' for <Function test_second_redis>>

pytest_redis/factories/client.py:61: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../.local/share/virtualenvs/pytest-redis-Dcbs60a9/lib/python3.13/site-packages/redis/commands/core.py:1559: in flushall
    return self.execute_command("FLUSHALL", *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        args       = []
        asynchronous = False
        kwargs     = {}
        self       = <redis.client.Redis(<redis.connection.ConnectionPool(<redis.connection.UnixDomainSocketConnection(db=0,username=<REDAC...elaxed_timeout=10, endpoint_type=None),orig_host_address=None,orig_socket_timeout=5,orig_socket_connect_timeout=5)>)>)>
../../../.local/share/virtualenvs/pytest-redis-Dcbs60a9/lib/python3.13/site-packages/redis/client.py:787: in execute_command
    return self._execute_command(*args, **options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        args       = ('FLUSHALL',)
        options    = {}
        self       = <redis.client.Redis(<redis.connection.ConnectionPool(<redis.connection.UnixDomainSocketConnection(db=0,username=<REDAC...elaxed_timeout=10, endpoint_type=None),orig_host_address=None,orig_socket_timeout=5,orig_socket_connect_timeout=5)>)>)>
../../../.local/share/virtualenvs/pytest-redis-Dcbs60a9/lib/python3.13/site-packages/redis/client.py:793: in _execute_command
    conn = self.connection or pool.get_connection()
                              ^^^^^^^^^^^^^^^^^^^^^
        args       = ('FLUSHALL',)
        command_name = 'FLUSHALL'
        options    = {}
        pool       = <redis.connection.ConnectionPool(<redis.connection.UnixDomainSocketConnection(db=0,username=<REDACTED>,*** relaxed_timeout=10, endpoint_type=None),orig_host_address=None,orig_socket_timeout=5,orig_socket_connect_timeout=5)>)>
        self       = <redis.client.Redis(<redis.connection.ConnectionPool(<redis.connection.UnixDomainSocketConnection(db=0,username=<REDAC...elaxed_timeout=10, endpoint_type=None),orig_host_address=None,orig_socket_timeout=5,orig_socket_connect_timeout=5)>)>)>
../../../.local/share/virtualenvs/pytest-redis-Dcbs60a9/lib/python3.13/site-packages/redis/utils.py:258: in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
        _check_deprecated_args = <function deprecated_args.<locals>._check_deprecated_args at 0x7f3b42376c00>
        allowed_args = []
        args       = (<redis.connection.ConnectionPool(<redis.connection.UnixDomainSocketConnection(db=0,username=<REDACTED>,*** endpoint_type=None),orig_host_address=None,orig_socket_timeout=5,orig_socket_connect_timeout=5)>)>,)
        filterable_args = {}
        func       = <function ConnectionPool.get_connection at 0x7f3b42376d40>
        kwargs     = {}
../../../.local/share/virtualenvs/pytest-redis-Dcbs60a9/lib/python3.13/site-packages/redis/connection.py:3139: in get_connection
    connection.connect()
        command_name = None
        connection = <redis.connection.UnixDomainSocketConnection(path=/tmp/pytest-of-runner/pytest-0/pytest-redis-redis_other_proc0/redis.29995.sock,db=0)>
        is_created = True
        keys       = ()
        options    = {}
        pool_name  = 'unknown:6379_4db4d7ec'
        self       = <redis.connection.ConnectionPool(<redis.connection.UnixDomainSocketConnection(db=0,username=<REDACTED>,*** relaxed_timeout=10, endpoint_type=None),orig_host_address=None,orig_socket_timeout=5,orig_socket_connect_timeout=5)>)>
        start_time_created = 66.298279715
../../../.local/share/virtualenvs/pytest-redis-Dcbs60a9/lib/python3.13/site-packages/redis/connection.py:1006: in connect
    self.retry.call_with_retry(
        self       = <redis.connection.UnixDomainSocketConnection(path=/tmp/pytest-of-runner/pytest-0/pytest-redis-redis_other_proc0/redis.29995.sock,db=0)>
../../../.local/share/virtualenvs/pytest-redis-Dcbs60a9/lib/python3.13/site-packages/redis/retry.py:120: in call_with_retry
    return do()
           ^^^^
        do         = <function AbstractConnection.connect.<locals>.<lambda> at 0x7f3b427994e0>
        fail       = <function AbstractConnection.connect.<locals>.<lambda> at 0x7f3b41db0860>
        failures   = 0
        is_retryable = None
        self       = <redis.retry.Retry object at 0x7f3b41d5b1d0>
        with_failure_count = False
../../../.local/share/virtualenvs/pytest-redis-Dcbs60a9/lib/python3.13/site-packages/redis/connection.py:1007: in <lambda>
    lambda: self.connect_check_health(
        self       = <redis.connection.UnixDomainSocketConnection(path=/tmp/pytest-of-runner/pytest-0/pytest-redis-redis_other_proc0/redis.29995.sock,db=0)>
../../../.local/share/virtualenvs/pytest-redis-Dcbs60a9/lib/python3.13/site-packages/redis/connection.py:1061: in connect_check_health
    self.on_connect_check_health(check_health=check_health)
        actual_retry_attempts = [0]
        check_health = True
        failure_callback = <function AbstractConnection.connect_check_health.<locals>.failure_callback at 0x7f3b41db13a0>
        retry_socket_connect = False
        self       = <redis.connection.UnixDomainSocketConnection(path=/tmp/pytest-of-runner/pytest-0/pytest-redis-redis_other_proc0/redis.29995.sock,db=0)>
        sock       = <socket.socket fd=11, family=1, type=1, proto=0, raddr=/tmp/pytest-of-runner/pytest-0/pytest-redis-redis_other_proc0/redis.29995.sock>
../../../.local/share/virtualenvs/pytest-redis-Dcbs60a9/lib/python3.13/site-packages/redis/connection.py:1163: in on_connect_check_health
    self.activate_maint_notifications_handling_if_enabled(check_health=check_health)
        auth_args  = None
        check_health = True
        parser     = <redis._parsers.resp3._RESP3Parser object at 0x7f3b41d32650>
        self       = <redis.connection.UnixDomainSocketConnection(path=/tmp/pytest-of-runner/pytest-0/pytest-redis-redis_other_proc0/redis.29995.sock,db=0)>
../../../.local/share/virtualenvs/pytest-redis-Dcbs60a9/lib/python3.13/site-packages/redis/connection.py:609: in activate_maint_notifications_handling_if_enabled
    self._enable_maintenance_notifications(
        check_health = True
        self       = <redis.connection.UnixDomainSocketConnection(path=/tmp/pytest-of-runner/pytest-0/pytest-redis-redis_other_proc0/redis.29995.sock,db=0)>
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <redis.connection.UnixDomainSocketConnection(path=/tmp/pytest-of-runner/pytest-0/pytest-redis-redis_other_proc0/redis.29995.sock,db=0)>
maint_notifications_config = MaintNotificationsConfig(enabled=auto, proactive_reconnect=True, relaxed_timeout=10, endpoint_type=None)
check_health = True

    def _enable_maintenance_notifications(
        self, maint_notifications_config: MaintNotificationsConfig, check_health=True
    ):
        try:
            host = getattr(self, "host", None)
            if host is None:
>               raise ValueError(
                    "Cannot enable maintenance notifications for connection"
                    " object that doesn't have a host attribute."
                )
E               ValueError: Cannot enable maintenance notifications for connection object that doesn't have a host attribute.

check_health = True
host       = None
maint_notifications_config = MaintNotificationsConfig(enabled=auto, proactive_reconnect=True, relaxed_timeout=10, endpoint_type=None)
self       = <redis.connection.UnixDomainSocketConnection(path=/tmp/pytest-of-runner/pytest-0/pytest-redis-redis_other_proc0/redis.29995.sock,db=0)>

../../../.local/share/virtualenvs/pytest-redis-Dcbs60a9/lib/python3.13/site-packages/redis/connection.py:620: ValueError

```

--- a/redis/client.py
+++ b/redis/client.py
@@ -398,10 +398,21 @@ class Redis(RedisModuleCommands, CoreCommands, SentinelCommands):
             }
             # based on input, setup appropriate connection args
             if unix_socket_path is not None:
+                if (
+                    maint_notifications_config
+                    and maint_notifications_config.enabled is True
+                ):
+                    raise RedisError(
+                        "Maintenance notifications are not supported with Unix "
+                        "domain socket connections"
+                    )
                 kwargs.update(
                     {
                         "path": unix_socket_path,
                         "connection_class": UnixDomainSocketConnection,
+                        "maint_notifications_config": MaintNotificationsConfig(
+                            enabled=False
+                        ),
                     }
                 )
             else:

--- a/redis/connection.py
+++ b/redis/connection.py
@@ -599,12 +599,13 @@ class MaintNotificationsAbstractConnection:
         # When the maint_notifications_config enabled mode is "auto",
         # we just log a warning if the handshake fails
         # When the mode is enabled=True, we raise an exception in case of failure
+        host = getattr(self, "host", None)
         if (
             self.get_protocol() not in [2, "2"]
             and self.maint_notifications_config
             and self.maint_notifications_config.enabled
             and self._maint_notifications_connection_handler
-            and hasattr(self, "host")
+            and host is not None
         ):
             self._enable_maintenance_notifications(
                 maint_notifications_config=self.maint_notifications_config,
@@ -2886,6 +2887,24 @@ class ConnectionPool(MaintNotificationsAbstractConnectionPool, ConnectionPoolInt
         self.max_connections = max_connections
         self.cache = None
         self._cache_factory = cache_factory
+
+        try:
+            supports_maint_notifications = issubclass(
+                connection_class, MaintNotificationsAbstractConnection
+            )
+        except TypeError:
+            supports_maint_notifications = False
+
+        if not supports_maint_notifications:
+            if (
+                maint_notifications_config
+                and maint_notifications_config.enabled is True
+            ):
+                raise RedisError(
+                    "Maintenance notifications are not supported with this "
+                    "connection class"
+                )
+            maint_notifications_config = MaintNotificationsConfig(enabled=False)
 
         self._event_dispatcher = self._connection_kwargs.get("event_dispatcher", None)
         if self._event_dispatcher is None:

--- a/redis/connection.py
+++ b/redis/connection.py
@@ -2892,17 +2892,21 @@ class ConnectionPool(MaintNotificationsAbstractConnectionPool, ConnectionPoolInt
             supports_maint_notifications = issubclass(
                 connection_class, MaintNotificationsAbstractConnection
             )
+            is_unix_domain_socket_connection = issubclass(
+                connection_class, UnixDomainSocketConnection
+            )
         except TypeError:
             supports_maint_notifications = False
+            is_unix_domain_socket_connection = False
 
-        if not supports_maint_notifications:
+        if is_unix_domain_socket_connection or not supports_maint_notifications:
             if (
                 maint_notifications_config
                 and maint_notifications_config.enabled is True
             ):
                 raise RedisError(
-                    "Maintenance notifications are not supported with this "
-                    "connection class"
+                    "Maintenance notifications are not supported with "
+                    f"{connection_class}"
                 )
             maint_notifications_config = MaintNotificationsConfig(enabled=False)
 

--- a/tests/test_connection_pool.py
+++ b/tests/test_connection_pool.py
@@ -659,17 +659,23 @@ class TestConnectionPoolUnixSocketURLParsing:
         ]
         socket_mock = mock.MagicMock()
 
-        with mock.patch.object(
-            redis.UnixDomainSocketConnection, "_connect", return_value=socket_mock
-        ), mock.patch.object(
-            redis.UnixDomainSocketConnection, "can_read", return_value=False
-        ), mock.patch.object(
-            redis.UnixDomainSocketConnection, "send_command"
-        ) as send_command, mock.patch.object(
-            redis.UnixDomainSocketConnection, "read_response", side_effect=responses
-        ), mock.patch.object(
-            redis.UnixDomainSocketConnection, "_enable_maintenance_notifications"
-        ) as enable:
+        with (
+            mock.patch.object(
+                redis.UnixDomainSocketConnection, "_connect", return_value=socket_mock
+            ),
+            mock.patch.object(
+                redis.UnixDomainSocketConnection, "can_read", return_value=False
+            ),
+            mock.patch.object(
+                redis.UnixDomainSocketConnection, "send_command"
+            ) as send_command,
+            mock.patch.object(
+                redis.UnixDomainSocketConnection, "read_response", side_effect=responses
+            ),
+            mock.patch.object(
+                redis.UnixDomainSocketConnection, "_enable_maintenance_notifications"
+            ) as enable,
+        ):
             assert client.ping() is True
             assert client.set(key, "value") is True
             assert client.get(key) == b"value"

--- a/tests/test_connection_pool.py
+++ b/tests/test_connection_pool.py
@@ -111,8 +111,7 @@ class TestConnectionPool:
         with pytest.raises(
             redis.RedisError,
             match=(
-                "Maintenance notifications are not supported with "
-                ".*DummyConnection"
+                "Maintenance notifications are not supported with .*DummyConnection"
             ),
         ):
             redis.ConnectionPool(

--- a/tests/test_connection_pool.py
+++ b/tests/test_connection_pool.py
@@ -110,7 +110,10 @@ class TestConnectionPool:
     def test_custom_connection_rejects_enabled_maint_notifications_config(self):
         with pytest.raises(
             redis.RedisError,
-            match="Maintenance notifications are not supported with this connection class",
+            match=(
+                "Maintenance notifications are not supported with "
+                ".*DummyConnection"
+            ),
         ):
             redis.ConnectionPool(
                 connection_class=DummyConnection,
@@ -645,6 +648,34 @@ class TestConnectionPoolUnixSocketURLParsing:
                 unix_socket_path="/socket",
                 maint_notifications_config=MaintNotificationsConfig(enabled=True),
             )
+
+    def test_pool_rejects_enabled_maint_notifications_config(self):
+        with pytest.raises(
+            redis.RedisError,
+            match=(
+                "Maintenance notifications are not supported with "
+                ".*UnixDomainSocketConnection"
+            ),
+        ):
+            redis.ConnectionPool(
+                connection_class=redis.UnixDomainSocketConnection,
+                path="/socket",
+                maint_notifications_config=MaintNotificationsConfig(enabled=True),
+            )
+
+    def test_pool_disables_default_maint_notifications(self):
+        pool = redis.ConnectionPool(
+            connection_class=redis.UnixDomainSocketConnection,
+            path="/socket",
+            protocol=3,
+        )
+
+        assert pool.connection_class == redis.UnixDomainSocketConnection
+        assert_kwargs_subset(pool.connection_kwargs, {"path": "/socket", "protocol": 3})
+        assert pool.maint_notifications_enabled() is None
+        assert pool._maint_notifications_pool_handler is None
+        assert "maint_notifications_config" not in pool.connection_kwargs
+        assert "maint_notifications_pool_handler" not in pool.connection_kwargs
 
     def test_default_config_client_executes_commands_without_maint_notifications(self):
         client = redis.Redis(unix_socket_path="/socket")

--- a/tests/test_connection_pool.py
+++ b/tests/test_connection_pool.py
@@ -100,6 +100,23 @@ class TestConnectionPool:
         assert isinstance(connection, DummyConnection)
         assert_kwargs_subset(connection.kwargs, connection_kwargs)
 
+    def test_custom_connection_disables_maint_notifications(self):
+        pool = redis.ConnectionPool(connection_class=DummyConnection)
+
+        assert pool.maint_notifications_enabled() is None
+        assert "maint_notifications_config" not in pool.connection_kwargs
+        assert "maint_notifications_pool_handler" not in pool.connection_kwargs
+
+    def test_custom_connection_rejects_enabled_maint_notifications_config(self):
+        with pytest.raises(
+            redis.RedisError,
+            match="Maintenance notifications are not supported with this connection class",
+        ):
+            redis.ConnectionPool(
+                connection_class=DummyConnection,
+                maint_notifications_config=MaintNotificationsConfig(enabled=True),
+            )
+
     @pytest.mark.fixed_client
     def test_closing(self):
         connection_kwargs = {"foo": "bar", "biz": "baz"}
@@ -589,6 +606,94 @@ class TestConnectionPoolUnixSocketURLParsing:
         pool = redis.ConnectionPool.from_url("unix:///socket")
         assert pool.connection_class == redis.UnixDomainSocketConnection
         assert_kwargs_subset(pool.connection_kwargs, {"path": "/socket"})
+
+    def test_client_disables_maint_notifications(self):
+        client = redis.Redis(unix_socket_path="/socket")
+        pool = client.connection_pool
+
+        assert pool.connection_class == redis.UnixDomainSocketConnection
+        assert_kwargs_subset(pool.connection_kwargs, {"path": "/socket"})
+        assert pool.maint_notifications_enabled() is None
+        assert "maint_notifications_config" not in pool.connection_kwargs
+        assert "maint_notifications_pool_handler" not in pool.connection_kwargs
+
+    def test_client_respects_disabled_maint_notifications_config(self):
+        client = redis.Redis(
+            unix_socket_path="/socket",
+            maint_notifications_config=MaintNotificationsConfig(enabled=False),
+        )
+        pool = client.connection_pool
+
+        assert pool.connection_class == redis.UnixDomainSocketConnection
+        assert_kwargs_subset(pool.connection_kwargs, {"path": "/socket"})
+        assert pool.maint_notifications_enabled() is None
+        assert pool._maint_notifications_pool_handler is None
+        # The disabled config is consumed by ConnectionPool and should not be
+        # propagated to individual Unix socket connections.
+        assert "maint_notifications_config" not in pool.connection_kwargs
+        assert "maint_notifications_pool_handler" not in pool.connection_kwargs
+
+    def test_client_rejects_enabled_maint_notifications_config(self):
+        with pytest.raises(
+            redis.RedisError,
+            match=(
+                "Maintenance notifications are not supported with Unix "
+                "domain socket connections"
+            ),
+        ):
+            redis.Redis(
+                unix_socket_path="/socket",
+                maint_notifications_config=MaintNotificationsConfig(enabled=True),
+            )
+
+    def test_default_config_client_executes_commands_without_maint_notifications(self):
+        client = redis.Redis(unix_socket_path="/socket")
+        key = "redis-py:unix-socket-default-config"
+        responses = [
+            {"proto": 3},
+            b"OK",
+            b"OK",
+            b"PONG",
+            b"OK",
+            b"value",
+        ]
+        socket_mock = mock.MagicMock()
+
+        with mock.patch.object(
+            redis.UnixDomainSocketConnection, "_connect", return_value=socket_mock
+        ), mock.patch.object(
+            redis.UnixDomainSocketConnection, "can_read", return_value=False
+        ), mock.patch.object(
+            redis.UnixDomainSocketConnection, "send_command"
+        ) as send_command, mock.patch.object(
+            redis.UnixDomainSocketConnection, "read_response", side_effect=responses
+        ), mock.patch.object(
+            redis.UnixDomainSocketConnection, "_enable_maintenance_notifications"
+        ) as enable:
+            assert client.ping() is True
+            assert client.set(key, "value") is True
+            assert client.get(key) == b"value"
+
+        client.close()
+        enable.assert_not_called()
+        sent_commands = [command.args for command in send_command.call_args_list]
+        assert ("PING",) in sent_commands
+        assert ("SET", key, "value") in sent_commands
+        assert ("GET", key) in sent_commands
+        for command in sent_commands:
+            assert command[:2] != ("CLIENT", "MAINT_NOTIFICATIONS")
+
+    def test_connection_does_not_activate_maint_notifications(self):
+        pool = redis.ConnectionPool(
+            connection_class=redis.UnixDomainSocketConnection,
+            path="/socket",
+        )
+        conn = pool.make_connection()
+
+        with mock.patch.object(conn, "_enable_maintenance_notifications") as enable:
+            conn.activate_maint_notifications_handling_if_enabled()
+
+        enable.assert_not_called()
 
     @skip_if_server_version_lt("6.0.0")
     def test_username(self):

